### ISSUE TITLE
RFC: Provider Breaking Changes

### ISSUE THREAD

**Issue by jharley** [honeycombio/terraform-provider-honeycombio#96]
# Overview

Honeycomb has recently taken ownership of the Terraform Provider which was originally developed and maintained by community member Koenraad Verheyden.

Leading up to the first Honeycomb-owned release of the Provider there are three breaking changes we feel need to be made:

1. The `honeycombio_query` data source will be renamed to `honeycombio_query_specification`. Initially raised in #65.
1. The `honeycombio_board` resource will no longer support inline Query JSON and instead require that you create a `honeycombio_query` resource and pass the ID in the list of queries on the Board. Initially raised in #55.
1. The `honeycombio_trigger` resource will no longer support inline Query JSON and instead require that you create a `honeycombio_query` resource and pass the ID to the Trigger resource. Also raised in #55.

As the move to the Honeycomb-maintained version of the provider will require that you update the namespace used in your `required_providers` configuration from `kvrhdn` to `honeycombio` we hope that these changes will be more easily managed.

This informal RFC was written to offer transparency around these changes and give the provider's user community of the a chance to comment on the changes.

## honeycombio_query data source change

This datasource never allowed you to fetch an existing Query object, but instead helped you form a valid Query Specification in HCL. Renaming this to `honeycombio_query_specification` seems most accurate and the change should easily be handled by a simple ‘find and replace’.

## honeycombio_board resource change

The Provider currently supports providing one or more inline queries (as Query Spec JSON `query_json`) to a Board resource to create a Board. This was created before either of the Query API or Query Annotations API existed, but both customers and the field team are looking to be able to build Boards via Terraform made up of fully annotated queries managed in code.

Inline queries are problematic because providing the query JSON to the API makes a Query object for you which is then difficult to track in Terraform as it’s a “computed sometimes” attribute of the resource. The explicit creation of a Query resource also feels more declarative and Terraform-ish.

Boards with Terraform today look something like this:

```
resource “honeycombio_board” “myboard” {
    name  = "Test board managed by Terraform"
    style = "list"

    query {
      caption       = "test query"
      dataset       = var.dataset
      query_json    = data.honeycombio_query.test.json
    }
    query {
      caption        = "my verbose query"
      query_style    = "combo"
      dataset        = var.dataset
      // Inline query
      query_json  = <<EOH
{
    "time_range": 604800,
<truncated for brevity>
}
EOH
    }
}
```

The current Board resource implementation ignores the returned `query_id` completely.

### Proposed change

The change to the Board resource’s schema will remove the ability to provide a Query Specification directly and instead require that you make use of the existing Query resource to construct Queries – with the recommendation that you do so with the assistance of the (newly renamed) Query Specification data source.

The above would turn into something like this:

```
resource “honeycombio_query” “helpful-query” {
   dataset       = var.dataset
   query_json    = data.honeycombio_query_specification.helpful.json
}

resource “honeycombio_query” “other-query” {
   dataset     = var.dataset
   query_json  = data.honeycombio_query_specification.other.json
}

resource “honeycombio_board” “myboard” {
    name  = "Test board managed by Terraform"
    style   = "list"

    query {
      caption       = "A very Helpful Query"
      dataset       = var.dataset
      query_id      = honeycombio_query.helpful-query.id
    }
    query {
      caption        = "my verbose query"
      query_style    = "table"
      dataset        = var.dataset
      query_id       = honeycombio_query.other-query.id
    }
}
```

The above paves the way for cleanly making use of the Query Annotation resource to name and describe the queries on the Board.

### Abandoned Ideas

Adding an “inline_query” type to the Board resource which allowed you to provide the Query Spec as JSON and would track the returned “query_id” as a “Calculated” schema attribute. Abandoned as it was still a breaking change (all current “query” would need to be moved to “inline_query”)
Attempting to use the Terraform Plugin SDK’s experimental “GetRawConfig” to extract the parsed HCL and determine if “query_json” was in use. This was ugly and felt like the sort of brittle that it would break in a future version of the Provider SDK.

## honeycombio_trigger resource change

Similar to the Board resource, the Provider currently supports passing an inline Query (as Query Specification JSON) to create a Trigger. The explicit creation of a Query resource also feels more declarative and Terraform-ish.

This change is ultimately intended to bring more consistency to the Provider’s interface.

Triggers in Terraform today look something like this:

```
resource “honeycombio_trigger” “mytrigger” {
  name       = "Trigger managed by Terraform"
  dataset    = var.dataset
  query_json = data.honeycombio_query.mytrigger.json

  frequency = 300

  threshold {
    op    = ">"
    value = 50
  }
  
  recipient {
    type   = "email"
    target = "hello@example.com"
  }
}
```

### Proposed change

The change to the Trigger resource’s schema will remove the ability to provide a Query Specification directly and instead require that you make use of the existing Query resource to construct Queries – with the recommendation that you do so with the assistance of the (newly renamed) Query Specification data source.

The above would turn into something like this:

```
resource “honeycombio_query” “mytrigger” {
  dataset    = var.dataset
  query_json = data.honeycombio_query_specification.mytrigger.json
}

resource “honeycombio_trigger” “mytrigger” {
  name     = "Trigger managed by Terraform"
  dataset  = var.dataset
  query_id = honeycombio_query.mytrigger.id

  frequency = 300

  threshold {
    op    = ">"
    value = 50
  }
  
  recipient {
    type   = "email"
    target = "hello@example.com"
  }
}
```

**Comment by slizco** [comment]
This makes a lot of sense to me. My only concern is around how the provider will handle management of the existing state with these resource changes/additions. I believe that making these changes in the Terraform code followed by an apply would simply show:
- newly created `honeycombio_query_specification` resources
- potential changes to `honeycombio_board` (particularly if query titles are added)
- no changes to `honeycombio_trigger` resources since the query definitions themselves will remain the same

In other words, I _think_ the transition will be seamless. Does that track with your expectations @jharley? 
I'd rather avoid having to do manual imports of resources.

**Comment by jharley** [comment]
@slizco Correct! These are just breaking changes to the schema which should not cause entire resources to be removed from state or put you into a position of needing to do 'state surgery' (manual imports or `state rm`)

**Comment by jharley** [comment]
@slizco Wait, I just thought about it a bit more: `honeycombio_trigger` resources will show an update as the Query ID will be changed but it should be an update in place.

**Comment by jharley** [comment]
Resolved by #98 and #100
